rasa/core/policies/ted_policy.py

```diff
@@ -654,264 +720,184 @@ def load(cls, path: Text) -> "TEDPolicy":
-        ):
+        if LABEL not in self.data_signature:
             raise ValueError(
                 f"No label features specified. "
                 f"Cannot train '{self.__class__.__name__}' model."
             )
 
-    def _prepare_sparse_dense_layers(
-        self,
-        data_signature: List[FeatureSignature],
-        name: Text,
-        reg_lambda: float,
-        dense_dim: int,
-    ) -> None:
-        sparse = False
-        dense = False
-        for is_sparse, shape in data_signature:
-            if is_sparse:
-                sparse = True
-            else:
-                dense = True
-
-        if sparse:
-            self._tf_layers[f"sparse_to_dense.{name}"] = layers.DenseForSparse(
-                units=dense_dim, reg_lambda=reg_lambda, name=name
-            )
-            if not dense:
-                # create dense labels for the input to use in negative sampling
-                self._tf_layers[f"sparse_to_dense_ids.{name}"] = layers.DenseForSparse(
-                    units=2, trainable=False, name=f"sparse_to_dense_ids.{name}"
-                )
-
     def _prepare_layers(self) -> None:
-        self._tf_layers[f"loss.{LABEL}"] = layers.DotProductLoss(
-            self.config[NUM_NEG],
-            self.config[LOSS_TYPE],
-            self.config[MAX_POS_SIM],
-            self.config[MAX_NEG_SIM],
-            self.config[USE_MAX_NEG_SIM],
-            self.config[NEGATIVE_MARGIN_SCALE],
-            self.config[SCALE_LOSS],
-            # set to 1 to get deterministic behaviour
-            parallel_iterations=1 if self.random_seed is not None else 1000,
-        )
-
-        for feature_name in self.data_signature.keys():
-            if feature_name.startswith(DIALOGUE_FEATURES) or feature_name.startswith(
-                LABEL_FEATURES
-            ):
-                self._prepare_sparse_dense_layers(
-                    self.data_signature[feature_name],
-                    feature_name,
-                    self.config[REGULARIZATION_CONSTANT],
-                    100,
-                )
+        self._prepare_dot_product_loss(LABEL, self.config[SCALE_LOSS])
 
-            if feature_name.startswith(DIALOGUE_FEATURES) and not feature_name.endswith(
-                "if_text"
-            ):
-                self._tf_layers[f"ffnn.{feature_name}"] = layers.Ffnn(
-                    self.config[HIDDEN_LAYERS_SIZES][f"{DIALOGUE}_name_text"],
-                    self.config[DROP_RATE_DIALOGUE],
-                    self.config[REGULARIZATION_CONSTANT],
-                    self.config[WEIGHT_SPARSITY],
-                    layer_name_suffix=feature_name,
-                )
+        for name in self.data_signature.keys():
+            self._prepare_utterance_level_layers(name)
 
-        self._tf_layers[f"ffnn.{DIALOGUE}"] = layers.Ffnn(
+        for name in FEATURES_TO_ENCODE:
+            self._prepare_encoding_layers(name)
+
+        self._prepare_ffnn_layer(
+            DIALOGUE,
```

Comment:
what is this ffnn layer?

Reply:
it's the FFNN layer before the transformer for all the dialogue features, e.g. action_text, intent, etc.

Reply:
aaahhh, the one used on line 924? could you please add explanatory comment

Reply:
done